File: frontend/src/tests/factory.ts
import dayjs from 'dayjs'
import type {
  User,
  Room,
  Booking,
  MeetingDoc,
  Todo,
  CheckIn,
  Notification,
  QRCodeData,
  RoomUsageStat,
  MeetingHoursStat,
  AttendanceStat,
  HeatmapData,
  EfficiencyStat,
  DisplayInfo,
} from '@/types'

const UUID = () => 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
  const r = (Math.random() * 16) | 0
  const v = c === 'x' ? r : (r & 0x3) | 0x8
  return v.toString(16)
})

export const uuid = UUID

export interface UserFactoryOptions {
  name?: string
  email?: string
  role?: string
  department?: string
}

export const createMockUser = (opts: UserFactoryOptions = {}): User => {
  const id = UUID()
  return {
    id,
    name: opts.name ?? `测试用户_${id.slice(0, 6)}`,
    email: opts.email ?? `user_${id.slice(0, 6)}@example.com`,
    phone: '13800000000',
    department: opts.department ?? '研发部',
    avatar: '',
    role: opts.role ?? 'user',
    created_at: dayjs().toISOString(),
    updated_at: dayjs().toISOString(),
  }
}

export const createMockAdmin = (overrides: Partial<User> = {}): User => ({
  ...createMockUser({ name: '系统管理员', email: 'admin@example.com', role: 'admin' }),
  ...overrides,
})

export interface RoomFactoryOptions {
  name?: string
  floor?: number
  capacity?: number
  status?: string
  need_approval?: boolean
}

export const createMockRoom = (opts: RoomFactoryOptions = {}): Room => {
  const id = UUID()
  return {
    id,
    name: opts.name ?? `会议室_${id.slice(0, 6)}`,
    floor: opts.floor ?? 1,
    capacity: opts.capacity ?? 10,
    equipment: '投影仪,白板,视频会议',
    description: '设备齐全的会议室',
    status: opts.status ?? 'active',
    need_approval: opts.need_approval ?? false,
    location: 'A座101',
    created_at: dayjs().toISOString(),
    updated_at: dayjs().toISOString(),
  }
}

export interface BookingFactoryOptions {
  room_id?: string
  user_id?: string
  title?: string
  start_time?: dayjs.Dayjs
  duration_minutes?: number
  status?: string
  approval_status?: string
  recurring_rule?: string
  recurring_id?: string
  room?: Room
  user?: User
}

export const createMockBooking = (opts: BookingFactoryOptions = {}): Booking => {
  const id = UUID()
  const start = opts.start_time ?? dayjs().add(1, 'day').hour(10).minute(0).second(0)
  const duration = opts.duration_minutes ?? 60
  const end = start.add(duration, 'minute')
  return {
    id,
    room_id: opts.room_id ?? UUID(),
    user_id: opts.user_id ?? UUID(),
    title: opts.title ?? `测试会议_${id.slice(0, 6)}`,
    description: '测试预订描述',
    start_time: start.format('YYYY-MM-DD HH:mm:ss'),
    end_time: end.format('YYYY-MM-DD HH:mm:ss'),
    status: opts.status ?? 'confirmed',
    recurring_rule: opts.recurring_rule,
    recurring_id: opts.recurring_id,
    approval_status: opts.approval_status ?? 'approved',
    room: opts.room,
    user: opts.user,
    created_at: dayjs().toISOString(),
    updated_at: dayjs().toISOString(),
  }
}

export const createMockRecurringBookings = (
  count: number,
  room: Room,
  user: User,
  base: dayjs.Dayjs = dayjs().add(1, 'day').hour(14).minute(0),
): Booking[] => {
  const recurring_id = UUID()
  return Array.from({ length: count }).map((_, i) =>
    createMockBooking({
      room_id: room.id,
      user_id: user.id,
      start_time: base.add(i * 7, 'day'),
      duration_minutes: 90,
      recurring_rule: 'weekly',
      recurring_id,
      title: `每周例会 #${i + 1}`,
      room,
      user,
    }),
  )
}

export interface MeetingDocFactoryOptions {
  booking_id?: string
  content?: string
  agenda?: string
  summary?: string
  is_archived?: boolean
}

export const createMockMeetingDoc = (opts: MeetingDocFactoryOptions = {}): MeetingDoc => {
  const id = UUID()
  return {
    id,
    booking_id: opts.booking_id ?? UUID(),
    agenda: opts.agenda ?? '## 会议议程\n- 项目进度\n- 问题讨论',
    content: opts.content ?? '讨论内容：项目进展顺利',
    summary: opts.summary,
    is_archived: opts.is_archived ?? false,
    created_at: dayjs().toISOString(),
    updated_at: dayjs().toISOString(),
  }
}

export const createMockContentWithTodos = (todos: string[]): string => {
  const items = todos.map((t, i) => `- [ ] ${t}`).join('\n')
  return `本次会议讨论内容\n\n${items}\n\n讨论完毕`
}

export interface TodoFactoryOptions {
  doc_id?: string
  booking_id?: string
  assignee_id?: string
  content?: string
  status?: string
  priority?: number
}

export const createMockTodo = (opts: TodoFactoryOptions = {}): Todo => {
  const id = UUID()
  return {
    id,
    doc_id: opts.doc_id ?? UUID(),
    booking_id: opts.booking_id ?? UUID(),
    content: opts.content ?? `待办事项_${id.slice(0, 6)}`,
    assignee_id: opts.assignee_id ?? UUID(),
    status: opts.status ?? 'pending',
    priority: opts.priority ?? 1,
    created_at: dayjs().toISOString(),
    updated_at: dayjs().toISOString(),
  }
}

export interface CheckInFactoryOptions {
  booking_id?: string
  user_id?: string
  status?: string
  offset_minutes?: number
}

export const createMockCheckIn = (opts: CheckInFactoryOptions = {}): CheckIn => {
  const id = UUID()
  return {
    id,
    booking_id: opts.booking_id ?? UUID(),
    user_id: opts.user_id ?? UUID(),
    check_in_at: dayjs().add(opts.offset_minutes ?? 0, 'minute').toISOString(),
    qr_code: UUID(),
    status: opts.status ?? 'checked_in',
    created_at: dayjs().toISOString(),
  }
}

export interface QRCodeDataOptions {
  booking_id?: string
  expires_offset_minutes?: number
}

export const createMockQRCodeData = (opts: QRCodeDataOptions = {}): QRCodeData => ({
  token: UUID() + UUID(),
  expires_at: dayjs()
    .add(opts.expires_offset_minutes ?? 5, 'minute')
    .toISOString(),
  booking_id: opts.booking_id ?? UUID(),
})

export interface NotificationFactoryOptions {
  user_id?: string
  type?: string
  title?: string
  content?: string
  status?: string
  booking_id?: string
}

export const createMockNotification = (opts: NotificationFactoryOptions = {}): Notification => {
  const id = UUID()
  return {
    id,
    user_id: opts.user_id ?? UUID(),
    type: opts.type ?? 'booking_confirm',
    title: opts.title ?? `通知_${id.slice(0, 6)}`,
    content: opts.content ?? '通知内容',
    channels: 'wechat,email',
    status: opts.status ?? 'unread',
    booking_id: opts.booking_id,
    created_at: dayjs().toISOString(),
  }
}

export const createMockRoomUsageStats = (rooms: Room[]): RoomUsageStat[] =>
  rooms.map((r, i) => ({
    room_id: r.id,
    room_name: r.name,
    total_hours: 5 + i * 2,
    usage_rate: 10 + i * 5,
    booking_count: 3 + i,
  }))

export const createMockMeetingHoursStats = (): MeetingHoursStat[] => [
  { department: '研发部', total_meetings: 45, total_hours: 80.5, avg_hours: 1.79 },
  { department: '产品部', total_meetings: 32, total_hours: 52.0, avg_hours: 1.63 },
  { department: '设计部', total_meetings: 18, total_hours: 27.0, avg_hours: 1.5 },
]

export const createMockHeatmapData = (): HeatmapData[] => {
  const data: HeatmapData[] = []
  for (let day = 0; day < 7; day++) {
    for (let hour = 8; hour < 20; hour++) {
      data.push({
        day_of_week: day,
        hour,
        count: day >= 1 && day <= 5 && hour >= 10 && hour <= 17 ? Math.floor(Math.random() * 6) : 0,
      })
    }
  }
  return data
}

export const createMockEfficiencyStats = (bookings: Booking[]): EfficiencyStat[] =>
  bookings.map((b) => {
    const planned = dayjs(b.end_time).diff(dayjs(b.start_time), 'minute')
    const actual = Math.round(planned * (0.6 + Math.random() * 0.5))
    return {
      booking_id: b.id,
      title: b.title,
      planned_minutes: planned,
      actual_minutes: actual,
      efficiency_rate: actual > 0 ? (actual / planned) * 100 : 0,
    }
  })

export const createMockDisplayInfo = (
  room: Room,
  current: Booking | null,
  next: Booking | null,
  bookings: Booking[],
): DisplayInfo => ({
  room,
  current_time: dayjs().toISOString(),
  current_booking: current ?? undefined,
  next_booking: next ?? undefined,
  today_bookings: bookings,
})

File: backend/tests/testutil/factory.go
package testutil

import (
	"meeting-system/internal/model"
	"testing"
	"time"

	"github.com/google/uuid"
	"github.com/stretchr/testify/require"
	"gorm.io/gorm"
)

type UserFactoryOption func(*model.User)

func WithRole(role string) UserFactoryOption {
	return func(u *model.User) {
		u.Role = role
	}
}

func WithDepartment(dept string) UserFactoryOption {
	return func(u *model.User) {
		u.Department = dept
	}
}

func WithName(name string) UserFactoryOption {
	return func(u *model.User) {
		u.Name = name
	}
}

func WithEmail(email string) UserFactoryOption {
	return func(u *model.User) {
		u.Email = email
	}
}

func CreateUser(t *testing.T, db *gorm.DB, opts ...UserFactoryOption) *model.User {
	t.Helper()
	id := uuid.New()
	user := &model.User{
		ID:           id,
		Name:         "测试用户_" + id.String()[:8],
		Email:        "test_" + id.String()[:8] + "@example.com",
		Phone:        "13800000000",
		Role:         "user",
		PasswordHash: "$2a$10$N9qo8uLOickgx2ZMRZoMyeIjZAgcfl7p92ldGxad68LJZdL17lhWy",
	}
	for _, opt := range opts {
		opt(user)
	}
	err := db.Create(user).Error
	require.NoError(t, err, "Failed to create user")
	return user
}

func CreateAdmin(t *testing.T, db *gorm.DB) *model.User {
	t.Helper()
	return CreateUser(t, db, WithRole("admin"), WithName("系统管理员"), WithEmail("admin@example.com"))
}

type RoomFactoryOption func(*model.Room)

func WithCapacity(capacity int) RoomFactoryOption {
	return func(r *model.Room) {
		r.Capacity = capacity
	}
}

func WithNeedApproval(need bool, approverID *uuid.UUID) RoomFactoryOption {
	return func(r *model.Room) {
		r.NeedApproval = need
		r.ApproverID = approverID
	}
}

func WithFloor(floor int) RoomFactoryOption {
	return func(r *model.Room) {
		r.Floor = floor
	}
}

func WithRoomStatus(status string) RoomFactoryOption {
	return func(r *model.Room) {
		r.Status = status
	}
}

func CreateRoom(t *testing.T, db *gorm.DB, opts ...RoomFactoryOption) *model.Room {
	t.Helper()
	id := uuid.New()
	room := &model.Room{
		ID:          id,
		Name:        "测试会议室_" + id.String()[:8],
		Floor:       1,
		Capacity:    10,
		Equipment:   "投影仪,白板,视频会议",
		Description: "测试用会议室",
		Status:      "active",
		Location:    "A座101",
	}
	for _, opt := range opts {
		opt(room)
	}
	err := db.Create(room).Error
	require.NoError(t, err, "Failed to create room")
	return room
}

type BookingFactoryOption func(*model.Booking)

func WithBookingStatus(status string) BookingFactoryOption {
	return func(b *model.Booking) {
		b.Status = status
	}
}

func WithApprovalStatus(status string) BookingFactoryOption {
	return func(b *model.Booking) {
		b.ApprovalStatus = status
	}
}

func WithRecurring(rule string, recurringID uuid.UUID) BookingFactoryOption {
	return func(b *model.Booking) {
		b.RecurringRule = rule
		b.RecurringID = &recurringID
	}
}

func WithTimeOffset(startOffset, endOffset time.Duration) BookingFactoryOption {
	return func(b *model.Booking) {
		b.StartTime = time.Now().Add(startOffset)
		b.EndTime = time.Now().Add(endOffset)
	}
}

func WithAttendees(attendees []string) BookingFactoryOption {
	return func(b *model.Booking) {
		if len(attendees) > 0 {
			first := attendees[0]
			b.Attendees = &first
		}
	}
}

func CreateBooking(t *testing.T, db *gorm.DB, roomID, userID uuid.UUID, startTime, endTime time.Time, opts ...BookingFactoryOption) *model.Booking {
	t.Helper()
	id := uuid.New()
	booking := &model.Booking{
		ID:             id,
		RoomID:         roomID,
		UserID:         userID,
		Title:          "测试会议_" + id.String()[:8],
		Description:    "测试用会议预订",
		StartTime:      startTime,
		EndTime:        endTime,
		Status:         "confirmed",
		ApprovalStatus: "approved",
	}
	for _, opt := range opts {
		opt(booking)
	}
	err := db.Create(booking).Error
	require.NoError(t, err, "Failed to create booking")
	return booking
}

func CreateMeetingDoc(t *testing.T, db *gorm.DB, bookingID uuid.UUID, content string) *model.MeetingDoc {
	t.Helper()
	doc := &model.MeetingDoc{
		ID:         uuid.New(),
		BookingID:  bookingID,
		Agenda:     "## 测试议程\n- 项目进度\n",
		Content:    content,
		Summary:    "",
		IsArchived: false,
	}
	err := db.Create(doc).Error
	require.NoError(t, err, "Failed to create meeting doc")
	return doc
}

func CreateTodo(t *testing.T, db *gorm.DB, docID, bookingID, assigneeID uuid.UUID, content string) *model.Todo {
	t.Helper()
	todo := &model.Todo{
		ID:         uuid.New(),
		DocID:      docID,
		BookingID:  bookingID,
		Content:    content,
		AssigneeID: assigneeID,
		Status:     "pending",
		Priority:   1,
	}
	err := db.Create(todo).Error
	require.NoError(t, err, "Failed to create todo")
	return todo
}

func CreateCheckIn(t *testing.T, db *gorm.DB, bookingID, userID uuid.UUID, qrToken string, opts ...func(*model.CheckIn)) *model.CheckIn {
	t.Helper()
	checkIn := &model.CheckIn{
		ID:        uuid.New(),
		BookingID: bookingID,
		UserID:    userID,
		CheckInAt: time.Now(),
		QRCode:    qrToken,
		Status:    "checked_in",
	}
	for _, opt := range opts {
		opt(checkIn)
	}
	err := db.Create(checkIn).Error
	require.NoError(t, err, "Failed to create check-in")
	return checkIn
}

func CreateQRToken(t *testing.T, db *gorm.DB, bookingID uuid.UUID, expiresAt time.Time) *model.QRCodeToken {
	t.Helper()
	token := &model.QRCodeToken{
		ID:        uuid.New(),
		BookingID: bookingID,
		Token:     uuid.New().String(),
		ExpiresAt: expiresAt,
	}
	err := db.Create(token).Error
	require.NoError(t, err, "Failed to create QR code token")
	return token
}

func CreateNotification(t *testing.T, db *gorm.DB, userID uuid.UUID, notifType, title, content string) *model.Notification {
	t.Helper()
	notif := &model.Notification{
		ID:       uuid.New(),
		UserID:   userID,
		Type:     notifType,
		Title:    title,
		Content:  content,
		Channels: "wechat,email",
		Status:   "unread",
	}
	err := db.Create(notif).Error
	require.NoError(t, err, "Failed to create notification")
	return notif
}

type QRTokenFactoryOption func(*model.QRCodeToken)

func WithExpiredToken() QRTokenFactoryOption {
	return func(q *model.QRCodeToken) {
		q.ExpiresAt = time.Now().Add(-10 * time.Minute)
	}
}

func (f *Factory) CreateQRCodeToken(bookingID uuid.UUID, opts ...QRTokenFactoryOption) *model.QRCodeToken {
	id := uuid.New()
	token := &model.QRCodeToken{
		ID:        id,
		BookingID: bookingID,
		Token:     "qr_token_" + id.String(),
		ExpiresAt: time.Now().Add(5 * time.Minute),
	}
	for _, opt := range opts {
		opt(token)
	}
	return token
}

type CheckInFactoryOption func(*model.CheckIn)

func WithCheckInTime(t time.Time) CheckInFactoryOption {
	return func(c *model.CheckIn) {
		c.CheckInAt = t
	}
}

func (f *Factory) CreateCheckIn(bookingID, userID uuid.UUID, opts ...CheckInFactoryOption) *model.CheckIn {
	id := uuid.New()
	checkIn := &model.CheckIn{
		ID:        id,
		BookingID: bookingID,
		UserID:    userID,
		CheckInAt: time.Now(),
		Status:    "checked_in",
	}
	for _, opt := range opts {
		opt(checkIn)
	}
	return checkIn
}

type Factory struct{}

func NewFactory() *Factory {
	return &Factory{}
}
